encoding.c

```diff
@@ -1564,15 +1619,21 @@ enc_set_default_encoding(struct default_encoding *def, VALUE encoding, const cha
 rb_encoding *
 rb_default_external_encoding(void)
 {
-    if (default_external.enc) return default_external.enc;
-
-    if (default_external.index >= 0) {
-        default_external.enc = rb_enc_from_index(default_external.index);
-        return default_external.enc;
-    }
-    else {
-        return rb_locale_encoding();
+    rb_encoding *enc = NULL;
+    // TODO: make lock-free
```

Comment:
What would it take to do this? I think we call these really a lot, and `rb_enc_from_index` only needs to lock in cases where the encoding is not one of the three fastpaths.

If it helps, I think it would be fine if we took a VM barrier to change internal or external encoding.

(ditto for internal encoding)